src/bitcoin.cpp

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+    return cmd;
+}
+
+void ExecCommand(const std::vector<std::string>& args, std::string_view argv0)
+{
+    // Construct argument string for execvp
+    std::vector<const char*> cstr_args{};
+    cstr_args.reserve(args.size() + 1);
+    for (const auto& arg : args) {
+        cstr_args.emplace_back(arg.c_str());
+    }
+    cstr_args.emplace_back(nullptr);
+
+    // Try to call execvp with given exe path.
+    auto try_exec = [&](fs::path exe_path, bool allow_notfound = true) {
+        std::string exe_path_str{fs::PathToString(exe_path)};
+        cstr_args[0] = exe_path_str.c_str();
+        if (execvp(cstr_args[0], (char*const*)cstr_args.data()) == -1) {
+            if (allow_notfound && errno == ENOENT) return false;
+            throw std::system_error(errno, std::system_category(), strprintf("execvp failed to execute '%s'", cstr_args[0]));
+        }
+        return true; // Will not actually be reached if execvp succeeds
+    };
+
+    // Try to figure out where current executable is located. This is a
+    // simplified search that won't work perfectly on every platform and doesn't
+    // need to, as it is only trying to prioritize locally built or installed
+    // executables over system executables. We may want to add options to
+    // override this behavior in the future, though.
+    fs::path exe_dir(argv0);
+    std::error_code ec;
+#ifndef WIN32
+    if (argv0.find('/') == std::string_view::npos) {
+        if (const char* path_env = std::getenv("PATH")) {
+            size_t start{0}, end{0};
+            for (std::string_view paths{path_env}; end != std::string_view::npos; start = end + 1) {
+                end = paths.find(':', start);
+                fs::path candidate = fs::path(paths.substr(start, end - start)) / exe_dir;
+                if (fs::exists(candidate, ec) && fs::is_regular_file(candidate, ec)) {
+                    exe_dir = candidate;
+                    break;
+                }
+            }
+        }
+    }
+#else
+    wchar_t module_path[MAX_PATH];
+    if (GetModuleFileNameW(nullptr, module_path, MAX_PATH) > 0) {
+        exe_dir = fs::path{module_path};
+    } else {
+        tfm::format(std::cerr, "Warning: Failed to get executable path. Error: %s\n", GetLastError());
+    }
+#endif
+    exe_dir = fs::weakly_canonical(exe_dir, ec).parent_path();
+
+    // If exe is in a CMake build tree, first look for target executable
+    // relative to it.
+    (exe_dir.filename() == "src" && try_exec(exe_dir / fs::PathFromString(args[0]))) ||
+    // Otherwise if exe is installed in a bin/ directory, first look for target
+    // executable in libexec/
+    (exe_dir.filename() == "bin" && try_exec(fs::path{exe_dir.parent_path()} / "libexec" / fs::PathFromString(args[0]).filename())) ||
+    // Otherwise look for target executable next to current exe
+    try_exec(exe_dir / fs::PathFromString(args[0]).filename()) ||
+    // Otherwise just look on the system path.
```

Comment:
b06c7ad0ae91102fe8cdcac6f86d627aace2219b: this is potentially confusing. If I build without gui, I expect `build/src/bitcoin gui` to fail. Right now it would launch any  `bitcoin-qt` in my `$PATH`.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31375#discussion_r1861814807

> [b06c7ad](https://github.com/bitcoin/bitcoin/commit/b06c7ad0ae91102fe8cdcac6f86d627aace2219b): this is potentially confusing. If I build without gui, I expect `build/src/bitcoin gui` to fail. Right now it would launch any `bitcoin-qt` in my `$PATH`.

Agree this behavior is potentially confusing, and prevented it for now, but I'm not totally I sure see it as a downside. I like the simplicity of `bitcoin daemon` being an alias for `bitcoind` and for the `bitcoin` wrapper to add a few additional search directories of its own but not otherwise care where other executables are installed on the PATH.

But for now added a check to avoid searching for other executables using the PATH if the `bitcoin` executable was not invoked using the PATH. Probably will need to keep tweaking this behavior, seeing what works and maybe adding options to control it.

Reply:
One advantage of allowing anything in `$PATH` would be once the interfaces are stable and "public", at which point folks might install e.g. an alternative GUI implementation.